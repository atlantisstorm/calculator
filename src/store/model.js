import {
  evaluateCalculation,
  parenthesesLeft,
  parenthesesRight,
  initialState
} from '../utils';

export const handleNumberClick = ({ state, number }) => {
  const { calculation, balancedParentheses } = state;
  let newState = state;
  let sequenceOK = true;
  let regex = /\.$/; // Ends with a dot.
  if (number.match(regex) && calculation.match(regex)) {
    sequenceOK = false;
  }
  regex = /\)$/; // Ends with a right parentheses.
  if (calculation.match(regex)) {
    sequenceOK = false;
  }
  if (number.match(/^\.$/)) { // clicked dot?
    // Split current calculation to ensure we're not adding a decimal point when we've already got one!
    const els = calculation.split(/[+\-*/]/);
    const last_els = els[els.length-1];
    if (last_els.match(/\./)) {
      sequenceOK = false;
    }
  }
  if (sequenceOK) {
    // The number pad key hit is the equals sign then we don't add to the calculation string.
    const newCalculation = number !== "=" ? `${calculation}${number}` : calculation;
    const { newResult, calculationError } = evaluateCalculation(newCalculation, balancedParentheses);
    newState = {
      ...state,
      calculationError,
      calculation: newCalculation,
      result: newResult
    };
  }
  return newState;
}

export const handleHistoryClick = ({ state }) => {
  const { calculation, result, history } = state;
  const newHistory = [...history, { calculation, result }];
  const newCalculation = "";
  const newResult = "";

  return {...state, calculation: newCalculation, result: newResult, history: newHistory };
}

export const handleOperatorClick = ({ state, operator }) => {
  // Values on operator panel ['+', '-', '*', '/'].
  const { calculation } = state;

  let newState = state;
  // Ensure we don't add multiple operators together nor that we're following a left parentheses with a divide/multipy.
  if (!(calculation.match(/[+\-*/]$/)) && !(calculation.match(/[(]$/) && operator.match(/[/*]/) )) {
    newState = {
      ...state,
      calculation: `${calculation}${operator}`
    };
  }

  return newState;
}

export const handleBackClick = ({ state }) => {
  const { calculation, balancedParentheses } = state;
  const lastCharacter = calculation.slice(-1); // get last character.
  const newCalculation = calculation.slice(0, -1); // Remove last character.

  let newBalancedParentheses = balancedParentheses;
  if (lastCharacter === parenthesesLeft) {
    newBalancedParentheses = balancedParentheses -1;
  } else if (lastCharacter === parenthesesRight) {
    newBalancedParentheses = balancedParentheses +1;
  }

  const { newResult } = evaluateCalculation(newCalculation, balancedParentheses);

  return {
    ...state,
    calculation: newCalculation,
    result: newResult,
    balancedParentheses: newBalancedParentheses
  };
}
  
export const handleClearClick = () => {
  return initialState;
}

export const handleParenthesesClick = ({ state, parentheses }) => {
  const { calculation, balancedParentheses } = state;
  let newCalculation = calculation;
  let newBalancedParentheses = balancedParentheses;

  if (parentheses === parenthesesLeft) {
    if (!calculation.match(/(\d|\))$/)) {
      newCalculation = `${calculation}${parenthesesLeft}`;
      newBalancedParentheses = balancedParentheses + 1;
    }
  } else {
    if (!calculation.match(/[+\-*(]$/) && balancedParentheses !== 0) {
      newCalculation = `${calculation}${parenthesesRight}`;
      newBalancedParentheses = balancedParentheses - 1;
    }
  }

  const { newResult, calculationError } = evaluateCalculation(newCalculation, newBalancedParentheses);

  return {
    ...state,
    calculationError,
    calculation: newCalculation,
    result: newResult,
    balancedParentheses: newBalancedParentheses
  };
}

export const Model = {
  handleNumberClick,
  handleHistoryClick,
  handleOperatorClick,
  handleBackClick,
  handleClearClick,
  handleParenthesesClick
};

export default Model;