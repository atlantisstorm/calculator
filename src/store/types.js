export const BACK_CLICKED = "BACK";
export const CLEAR_CLICKED = "CLEAR";
export const HISTORY_CLICKED = "HISTORY";
export const NUMBER_CLICKED = "NUMBER";
export const OPERATOR_CLICKED = "OPERATOR";
export const PARENTHESES_CLICKED = "PARENTHESES";