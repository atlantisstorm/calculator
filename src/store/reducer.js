import Model from './model';
import {
  BACK_CLICKED,
  CLEAR_CLICKED,
  HISTORY_CLICKED,
  NUMBER_CLICKED,
  OPERATOR_CLICKED,
  PARENTHESES_CLICKED
} from './types';

export const reducer = (state, action) => {
  switch (action.type) {
    case NUMBER_CLICKED: {
      // Values on number panel [7,8,9,4,5,6,1,2,3,'.',0, "="].
      return Model.handleNumberClick({
        state,
        number: action.payload
      });
    }

    case HISTORY_CLICKED: {
      return Model.handleHistoryClick({ state });
    }

    case OPERATOR_CLICKED: {
      // Values on operator panel ['+', '-', '*', '/'].
      return Model.handleOperatorClick({ 
        state,
        operator: action.payload
      });
    }

    case BACK_CLICKED: {
      return Model.handleBackClick({ state });
    }

    case CLEAR_CLICKED: {
      return Model.handleClearClick();
    }

    case PARENTHESES_CLICKED: {
      return Model.handleParenthesesClick({
        state,
        parentheses: action.payload
      })
    }

    default: {
      return state;
    }
  }
};