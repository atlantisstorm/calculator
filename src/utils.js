import { evaluate } from 'mathjs';
export const initialState = {
  calculation: "",
  result: "",
  balancedParentheses: 0,
  calculationError: false,
  history: []
}

export const numBtns = [7,8,9,4,5,6,1,2,3,'.',0, "="];
export const operatorsBtns = ['+', '-', '*', '/'];

export const parenthesesLeft = '(';
export const parenthesesRight = ')';

export const evaluateCalculation = (calculation, balancedParentheses) => {
  let newResult = "";
  let calculationError = false;
  try {
    newResult = balancedParentheses === 0 ? evaluate(calculation) : "";
  } catch(err) {
    calculationError = true;
  }

  if (newResult === Infinity || newResult === -Infinity) {
    newResult = "Cannot divide by 0";
  }

  return { newResult, calculationError };
}