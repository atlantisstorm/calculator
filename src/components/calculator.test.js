import React, { useReducer } from 'react';
import { render, fireEvent } from "@testing-library/react";
import Context from "../store/context";
import { reducer } from '../store/reducer';
import { initialState } from '../utils';
import Calculator from './calculator';

import {
  BACK_CLICKED,
  CLEAR_CLICKED,
  HISTORY_CLICKED,
  NUMBER_CLICKED,
  OPERATOR_CLICKED,
  PARENTHESES_CLICKED
} from '../store/types';

describe('Calculator component', () => {
  it('should match snapshot', () => {
    const TestComponent = () => {
      const [state, dispatch] = useReducer(reducer, initialState);
      return (
        <Context.Provider value={{ state, dispatch }}>
          <Calculator />
        </Context.Provider>
      )
    }

    const { asFragment } = render(
      <TestComponent />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should call dispatch with expected parameters', () => {
    const dispatchMock = jest.fn();
    let coreButton;

    const TestComponent = () => {
      const [state, dispatch] = useReducer(reducer, initialState);
      return (
        <Context.Provider value={{ state, dispatch: dispatchMock }}>
          <Calculator />
        </Context.Provider>
      )
    }

    const { container, asFragment, getByTestId } = render(
      <TestComponent />
    );
  
    expect(dispatchMock).toHaveBeenCalledTimes(0);

    // number panel click
    coreButton = getByTestId('core-button-1');
    fireEvent.click(coreButton);
    expect(dispatchMock).toHaveBeenCalledTimes(1);
    expect(dispatchMock).toHaveBeenCalledWith({ type: NUMBER_CLICKED, payload: "1" });

    // operator panel click
    coreButton = getByTestId('core-button-+');
    fireEvent.click(coreButton);
    expect(dispatchMock).toHaveBeenCalledTimes(2);
    expect(dispatchMock).toHaveBeenCalledWith({ type: OPERATOR_CLICKED, payload: "+" });

    // top panel
    coreButton = getByTestId('core-button-(');
    fireEvent.click(coreButton);
    expect(dispatchMock).toHaveBeenCalledTimes(3);
    expect(dispatchMock).toHaveBeenCalledWith({ type: PARENTHESES_CLICKED, payload: "(" });

    // history (=) button clicked
    coreButton = getByTestId('core-button-=');
    fireEvent.click(coreButton);
    expect(dispatchMock).toHaveBeenCalledTimes(4);
    expect(dispatchMock).toHaveBeenCalledWith({ type: HISTORY_CLICKED });

    // back button clicked
    coreButton = getByTestId('core-button-back');
    fireEvent.click(coreButton);
    expect(dispatchMock).toHaveBeenCalledTimes(5);
    expect(dispatchMock).toHaveBeenCalledWith({ type: BACK_CLICKED });

    // clear button clicked
    coreButton = getByTestId('core-button-clear');
    fireEvent.click(coreButton);
    expect(dispatchMock).toHaveBeenCalledTimes(6);
    expect(dispatchMock).toHaveBeenCalledWith({ type: CLEAR_CLICKED });

    // Plus loads of other tests...
  });

  it('should call dispatch with expected parameters', () => {
    let coreButton;  
    const TestComponent = () => {
      const [state, dispatch] = useReducer(reducer, initialState);
      return (
        <Context.Provider value={{ state, dispatch }}>
          <Calculator />
        </Context.Provider>
      )
    }

    const { container, getByTestId } = render(
      <TestComponent />
    );
 
    // click 5
    coreButton = getByTestId('core-button-5');
    fireEvent.click(coreButton);

    // click *
    coreButton = getByTestId('core-button-*');
    fireEvent.click(coreButton);

    // click 5
    coreButton = getByTestId('core-button-5');
    fireEvent.click(coreButton);

    // operator panel click
    let calculation = getByTestId('calculation');
    let result = getByTestId('result');

    expect(calculation.innerHTML).toContain("5*5");
    expect(result.innerHTML).toContain("25");

    // click back
    coreButton = getByTestId('core-button-back');
    fireEvent.click(coreButton);

    calculation = getByTestId('calculation');
    expect(calculation.innerHTML).toContain("5*");

    // clear click
    coreButton = getByTestId('core-button-clear');
    fireEvent.click(coreButton);

    calculation = getByTestId('calculation');
    result = getByTestId('result');
    expect(calculation.innerHTML).toContain("");
    expect(result.innerHTML).toContain("");    

    // Plus loads of other tests...
  });
});