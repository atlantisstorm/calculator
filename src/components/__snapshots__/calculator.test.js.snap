// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="panel"
    >
      <h4>
        Calculator
      </h4>
      <div>
        <div
          class="calculation"
          data-testid="calculation"
          placeholder="Results will appear here"
        />
      </div>
      <div
        class="result"
        data-testid="result"
      >
        <span />
      </div>
      <div>
        <button
          class="parentheses"
          data-testid="core-button-("
          value="("
        >
          (
        </button>
        <button
          class="parentheses"
          data-testid="core-button-)"
          value=")"
        >
          )
        </button>
        <button
          class="back"
          data-testid="core-button-back"
          value="back"
        >
          &lt;back&gt;
        </button>
      </div>
      <div>
        <div
          class="number-panel"
        >
          <button
            class="number"
            data-testid="core-button-7"
            value="7"
          >
            7
          </button>
          <button
            class="number"
            data-testid="core-button-8"
            value="8"
          >
            8
          </button>
          <button
            class="number"
            data-testid="core-button-9"
            value="9"
          >
            9
          </button>
          <button
            class="number"
            data-testid="core-button-4"
            value="4"
          >
            4
          </button>
          <button
            class="number"
            data-testid="core-button-5"
            value="5"
          >
            5
          </button>
          <button
            class="number"
            data-testid="core-button-6"
            value="6"
          >
            6
          </button>
          <button
            class="number"
            data-testid="core-button-1"
            value="1"
          >
            1
          </button>
          <button
            class="number"
            data-testid="core-button-2"
            value="2"
          >
            2
          </button>
          <button
            class="number"
            data-testid="core-button-3"
            value="3"
          >
            3
          </button>
          <button
            class="number"
            data-testid="core-button-."
            value="."
          >
            .
          </button>
          <button
            class="number"
            data-testid="core-button-0"
            value="0"
          >
            0
          </button>
          <button
            class="equals"
            data-testid="core-button-="
            value="="
          >
            =
          </button>
        </div>
        <div
          class="operators-panel"
        >
          <button
            class="operator"
            data-testid="core-button-+"
            value="+"
          >
            +
          </button>
          <button
            class="operator"
            data-testid="core-button--"
            value="-"
          >
            -
          </button>
          <button
            class="operator"
            data-testid="core-button-*"
            value="*"
          >
            *
          </button>
          <button
            class="operator"
            data-testid="core-button-/"
            value="/"
          >
            /
          </button>
        </div>
      </div>
      <div
        class="clear"
      />
      <div>
        <button
          class="clear"
          data-testid="core-button-clear"
          value="clear"
        >
          Clear
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
